From 141fbf8a11a456171a5fce773f01ffef4edf0016 Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Wed, 5 Dec 2018 22:34:02 +0000
Subject: [PATCH] OS-7395 bhyve: VM_{SUSPEND,RESUME}_CPU ioctls ignore vcpu
 argument Reviewed by: John Levon <john.levon@joyent.com> Reviewed by: Patrick
 Mooney <patrick.mooney@joyent.com> Approved by: John Levon
 <john.levon@joyent.com>

---
 usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
index 8185ca8ea5..2e92ff1624 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
@@ -969,10 +969,30 @@ vmmdev_do_ioctl(vmm_softc_t *sc, int cmd, intptr_t arg, int md,
 	case VM_ACTIVATE_CPU:
 		error = vm_activate_cpu(sc->vmm_vm, vcpu);
 		break;
+
 	case VM_SUSPEND_CPU:
+		if (ddi_copyin(datap, &vcpu, sizeof (vcpu), md)) {
+			error = EFAULT;
+			break;
+		}
+		if (vcpu < -1 || vcpu >= VM_MAXCPU) {
+			error = EINVAL;
+			break;
+		}
+
 		error = vm_suspend_cpu(sc->vmm_vm, vcpu);
 		break;
+
 	case VM_RESUME_CPU:
+		if (ddi_copyin(datap, &vcpu, sizeof (vcpu), md)) {
+			error = EFAULT;
+			break;
+		}
+		if (vcpu < -1 || vcpu >= VM_MAXCPU) {
+			error = EINVAL;
+			break;
+		}
+
 		error = vm_resume_cpu(sc->vmm_vm, vcpu);
 		break;
 
-- 
2.21.0

